micro-rdk/src/common/data_manager.rs

```diff
@@ -231,13 +247,68 @@ pub enum DataSyncError {
     DataStoreError(#[from] DataStoreError),
     #[error(transparent)]
     MessageDecodingError(#[from] prost::DecodeError),
+    #[error("time correction resulted in out of bounds duration")]
+    TimeOutOfBoundsError,
+    #[error("current time unset")]
+    NoCurrentTime,
 }
 
 pub struct DataSyncTask<StoreType> {
     store: Rc<AsyncMutex<StoreType>>,
     resource_method_keys: Vec<ResourceMethodKey>,
     sync_interval: Duration,
     part_id: String,
+    robot_start_time: Instant,
+}
+
+fn get_time_to_subtract(
+    robot_start_time: Instant,
+    stored_time: Timestamp,
+) -> Result<chrono::Duration, DataSyncError> {
+    let stored_time_dur = Duration::new(stored_time.seconds as u64, stored_time.nanos as u32);
+    let time_to_subtract = robot_start_time.elapsed() - stored_time_dur;
+    let time_to_subtract = chrono::Duration::new(
+        time_to_subtract.as_secs() as i64,
+        time_to_subtract.subsec_nanos(),
+    )
+    .ok_or(DataSyncError::TimeOutOfBoundsError)?;
+    Ok(time_to_subtract)
```

Comment:
Is there some from/into type magic you can use here to traverse the three types at play here (`Duration`, `Instant`, and `Timestamp`) so we can absolve ourselves of needing to get it right?

Reply:
I didn't see any, but I'll double-check

Reply:
Might as well hold off on any changes here until we have a resolution regarding `Instant`

Reply:
Now that things are settled on `Instant`, let's have another quick look at this (like no more than 10 minutes) to see if we can make the existing libraries do the work for us.

Reply:
Unfortunately, the answer appears to be no.